File: ios/DataLoader.m
// Copyright 2018 David Sansome
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#import "DataLoader.h"

#include <arpa/inet.h>

static uint32_t ReadUint32(NSFileHandle *file, size_t offset) {
  [file seekToFileOffset:offset];
  NSData *data = [file readDataOfLength:sizeof(uint32_t)];
  return *(uint32_t *)(data.bytes);
}

@implementation DataLoader {
  NSFileHandle *_file;
  uint32_t _firstSubjectOffset;
  TKMDataFileHeader *_header;
}

- (instancetype)initFromURL:(NSURL *)url {
  if (self = [super init]) {
    NSError *err = nil;
    _file = [NSFileHandle fileHandleForReadingFromURL:url error:&err];
    NSAssert(err == nil, @"Opening data file at %@ failed: %@", url, err);

    // Read the header.
    uint32_t headerLength = ReadUint32(_file, 0);
    _firstSubjectOffset = 4 + headerLength;
    NSData *headerData = [_file readDataOfLength:headerLength];
    _header = [TKMDataFileHeader parseFromData:headerData error:&err];
    NSAssert(err == nil, @"Reading data file header failed: %@", err);
  }
  return self;
}

- (int)maxSubjectLevel {
  return (int)_header.subjectsByLevelArray_Count;
}

- (GPBInt32Array *)deletedSubjectIDs {
  return _header.deletedSubjectIdsArray;
}

- (int)maxLevelGrantedBySubscription {
  if (_maxLevelGrantedBySubscription) {
    return MIN(_maxLevelGrantedBySubscription, self.maxSubjectLevel);
  }
  return self.maxSubjectLevel;
}

- (bool)isValidSubjectID:(int)subjectID {
  return subjectID < _header.subjectByteOffsetArray_Count && subjectID >= 0;
}

- (nullable TKMSubject *)loadSubject:(int)subjectID {
  NSAssert([self isValidSubjectID:subjectID], @"Tried to read subject %d outside 0-%d", subjectID,
           (int)_header.subjectByteOffsetArray_Count);

  const uint32_t offset =
      _firstSubjectOffset + [_header.subjectByteOffsetArray valueAtIndex:subjectID];

  NSData *data = nil;
  if (subjectID == _header.subjectByteOffsetArray_Count - 1) {
    [_file seekToFileOffset:offset];
    data = [_file readDataToEndOfFile];
  } else {
    // Read the offset of the next subject and compare to determine the length.
    const uint32_t nextOffset =
        _firstSubjectOffset + [_header.subjectByteOffsetArray valueAtIndex:subjectID + 1];
    const uint32_t length = nextOffset - offset;

    [_file seekToFileOffset:offset];
    data = [_file readDataOfLength:length];
  }

  TKMSubject *ret = [TKMSubject parseFromData:data error:nil];
  if (ret.level > self.maxLevelGrantedBySubscription) {
    NSLog(@"Tried to load subject %d from level %d > max level %d",
          subjectID, ret.level, self.maxLevelGrantedBySubscription);
    return nil;
  }
  ret.id_p = subjectID;
  return ret;
}

- (NSArray<TKMSubject *> *)loadAllSubjects {
  NSMutableArray<TKMSubject *> *ret = [NSMutableArray array];
  for (int subjectID = 1; subjectID < _header.subjectByteOffsetArray_Count; ++subjectID) {
    TKMSubject *subject = [self loadSubject:subjectID];
    if (subject) {
      [ret addObject:subject];
    }
  }
  return ret;
}

- (nullable TKMSubjectsByLevel *)subjectsByLevel:(int)level {
  if (level > self.maxLevelGrantedBySubscription) {
    return nil;
  }
  return _header.subjectsByLevelArray[level - 1];
}

@end

File: ios/DataLoader.swift
// Copyright 2021 David Sansome
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import Foundation

private func readUInt32(_ fh: FileHandle, _ offset: UInt64) -> UInt32 {
  fh.seek(toFileOffset: offset)
  let data = fh.readData(ofLength: 4)
  return data.withUnsafeBytes {
    (p: UnsafeRawBufferPointer) in
    p.bindMemory(to: UInt32.self).first!
  }
}

@objcMembers
class DataLoader: NSObject {
  let header: TKMDataFileHeader

  private let file: FileHandle
  private let firstSubjectOffset: UInt32

  public init(fromURL url: URL) throws {
    // Open the file.
    file = try FileHandle(forReadingFrom: url)

    // Read the header.
    let headerLength = readUInt32(file, 0)
    let headerData = file.readData(ofLength: Int(headerLength))
    header = try TKMDataFileHeader(data: headerData)

    // The subjects start after the header.
    firstSubjectOffset = 4 + headerLength
  }

  func levelOf(subjectID id: Int) -> Int {
    if id < 0 || id >= header.levelBySubjectArray.count {
      return 0
    }
    return Int(header.levelBySubjectArray.value(at: UInt(id)))
  }

  @objc(isValidSubjectID:)
  func isValid(subjectID id: Int) -> Bool {
    let level = levelOf(subjectID: id)
    return level > 0 && level <= maxLevelGrantedBySubscription
  }

  @objc(loadSubject:)
  func load(subjectID id: Int) -> TKMSubject? {
    if !isValid(subjectID: id) {
      return nil
    }

    do {
      let offset = firstSubjectOffset + header.subjectByteOffsetArray.value(at: UInt(id))
      file.seek(toFileOffset: UInt64(offset))

      var data: Data
      if id == header.subjectByteOffsetArray.count - 1 {
        data = file.readDataToEndOfFile()
      } else {
        // Read the offset of the next subject and compare to determine the length.
        let nextOffset = firstSubjectOffset + header.subjectByteOffsetArray.value(at: UInt(id + 1))
        let length = nextOffset - offset

        data = file.readData(ofLength: Int(length))
      }

      let ret = try TKMSubject(data: data)
      ret.id_p = Int32(id)
      return ret
    } catch {
      return nil
    }
  }

  var maxSubjectLevel: Int {
    header.subjectsByLevelArray.count
  }

  private var _maxLevelGrantedBySubscription = 60
  var maxLevelGrantedBySubscription: Int {
    get {
      min(_maxLevelGrantedBySubscription, maxSubjectLevel)
    }
    set {
      _maxLevelGrantedBySubscription = newValue
    }
  }

  func loadAll() -> [TKMSubject] {
    var ret = [TKMSubject]()
    for id in 1 ... header.subjectByteOffsetArray.count {
      if let subject = load(subjectID: Int(id)) {
        ret.append(subject)
      }
    }
    return ret
  }

  func subjects(byLevel level: Int) -> TKMSubjectsByLevel? {
    if level <= 0 || level > maxLevelGrantedBySubscription {
      return nil
    }

    return header.subjectsByLevelArray?[level - 1] as? TKMSubjectsByLevel
  }

  var deletedSubjectIDs: GPBInt32Array {
    header.deletedSubjectIdsArray
  }
}
